crates/tt/src/iter.rs

```diff
@@ -132,6 +132,7 @@ impl<'a, S: Copy> TtIter<'a, S> {
             }
             ('-' | '!' | '*' | '/' | '&' | '%' | '^' | '+' | '<' | '=' | '>' | '|', '=', _)
             | ('-' | '=' | '>', '>', _)
+            | (_, _, Some(';'))
```

Comment:
This is needed because we incorrectly treat ; as a possible glued token, when it shouldn't be as it's an expression terminator. It does fix the bug in the repro mentioned above.

Reply:
To clarify, why this works is that we do a 3-token at a time lookahead, and we see :%; which without this matches the rule for % followed by : followed by % (which is allowed because it's not forbidden by the match expression).

Reply:
I see, it is clear that this entire part is messy and needs to be redone at some point (likely by https://github.com/rust-lang/rust-analyzer/pull/17830) but given this improves things right now lets merge it, thanks!